### PR TITLE
Setting etcd3 as the default version for k8s 1.9+

### PR DESCRIPTION
**Pull request by chrislovecnm**
If the etcd version is not set, and we are using Kubernetes 1.8+ the
etcd version is set.  Currently, we have enabled etcd 3.0.17 for k8s
1.8.x, and etcd 3.1.10 for k8s 1.9.x+.

### CODE REVIEW

**Comment by roeera** [comment]
Hi,
Actually we are waiting to this feature also (mostly to get more monitoring capabilities on etcd).
The only question is that moving to etcd3 would be quite risky as first you need to perform a backup on etcd2.
Are you taking this in consideration ?

**Comment by chrislovecnm** [comment]
@roeera this PR does not support migration, that is a work in progress.

**Comment by justinsb** [comment]
I don't think we should change 1.8.  I'd like to get at least backup in place first.

Working on backup / upgrades here: https://github.com/kopeio/etcd-manager

File: upup/pkg/fi/cloudup/defaults.go [hunk 1]
```diff
@@ -227,3 +238,36 @@ func incrementIP(ip net.IP, cidr string) (string, error) {
 	}
 	return ip.String(), nil
 }
+
+// setupEctdVersion enables the supported Etcd version, if not set.
+func setupEctdVersion(cluster *kops.Cluster) error {
```
Comment (KashifSaadat) [inline upup/pkg/fi/cloudup/defaults.go]: Typo, correct to: `setupEtcdVersion`

**Comment by chrislovecnm** [comment]
/assign @justinsb 
/hold

I know you want backup and migration in at the same time. Please remove the hold and merge when you like!

**Comment by spiffxp** [comment]
/lgtm cancel
per the typo

/unassign

**Comment by iMartyn** [comment]
@justinsb could we consider pulling this forward into the 1.9 milestone? It's embarrassing to admit that by default kubernetes 1.9 installed with 1.9 won't scale properly (etcd2 bottlenecks at large clusters) and won't support secrets encrypted at rest.  As it's only a change in the default, it shouldn't really impact the timeline much if at all.

**Comment by vietwow** [comment]
I'm really need this in k8s 1.9 too.

**Comment by sstarcher** [comment]
I believe what is holding it back is migration tooling from etcd2 to etcd3.  That's my understanding why it has not moved to the default, but maybe we could default new clusters to etcd3 and not upgrade existing clusters?

File: upup/pkg/fi/cloudup/defaults.go [hunk 2]
```diff
@@ -227,3 +238,36 @@ func incrementIP(ip net.IP, cidr string) (string, error) {
...
+	if err != nil {
+		return fmt.Errorf("error parsing version: %v", cluster.Spec.KubernetesVersion)
+	}
+
+	version := ""
+	// as more versions are release we can add more if else statements here to set version.
+	if util.IsKubernetesGTE("1.9.0", *parsedVersion) {
+		// https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#external-dependencies
+		// lists 3.1.10 as the supported version for 1.9.x
+		version = "3.1.10"
```
Comment (KashifSaadat) [inline upup/pkg/fi/cloudup/defaults.go]: Could we default to the v3.3.x release as this includes quite an important fix for the `mvcc: database space exceeded` error: https://github.com/coreos/etcd/blob/master/CHANGELOG-3.3.md#v330-2018-02-01

File: upup/pkg/fi/cloudup/defaults.go [hunk 3]
```diff
@@ -227,3 +238,36 @@ func incrementIP(ip net.IP, cidr string) (string, error) {
...
+	// 	version = "3.0.17"
+	// }
+
+	if version == "" {
+		return nil
+	}
+
+	for _, etcd := range cluster.Spec.EtcdClusters {
+		if etcd.Version == "" {
+			etcd.Version = version
```
Comment (KashifSaadat) [inline upup/pkg/fi/cloudup/defaults.go]: If version is being updated as per previous comment, we would also need to set the image location:
```
etcd.Image = "gcr.io/etcd-development/etcd:v3.3.0"
```

Comment (KashifSaadat) [reply upup/pkg/fi/cloudup/defaults.go]: Also enable etcd TLS by default?

**Comment by justinsb** [comment]
This should be covered by https://github.com/kubernetes/kops/pull/5820

**Pull request closed without merging**